stdlib/public/TensorFlow/Ops.swift

```diff
@@ -1597,52 +1597,299 @@ public extension Tensor where Scalar : Numeric {
+    let value = slice(lowerBounds: lowerBounds, sizes: sizes)
+    let afterPaddings = shapeTensor - value.shapeTensor - lowerBounds
+    return (value, { [after = afterPaddings] v in
+      let beforePaddings = lowerBounds.expandingShape(at: 1)
+      let afterPaddings = after.expandingShape(at: 1)
+      let paddings = Tensor<Int32>(
+        concatenating: [beforePaddings, afterPaddings], alongAxis: 1)
+      return Raw.pad(v, paddings: paddings)
+    })
+  }
+}
+
+public enum TensorRange : TensorRangeExpression {
+  case ellipsis
+  case newAxis
+  case squeezeAxis
+  case index(Int32)
+  case range(Range<Int32>, stride: Int32)
+  case closedRange(ClosedRange<Int32>, stride: Int32)
+  case partialRangeFrom(PartialRangeFrom<Int32>, stride: Int32)
+  case partialRangeUpTo(PartialRangeUpTo<Int32>, stride: Int32)
+  case partialRangeThrough(PartialRangeThrough<Int32>, stride: Int32)
+
+  public var tensorRange: TensorRange { return self }
+}
+
+extension TensorRange : Equatable {
+  public static func == (lhs: TensorRange, rhs: TensorRange) -> Bool {
+    switch (lhs, rhs) {
+    case (.ellipsis, .ellipsis),
+         (.newAxis, .newAxis),
+         (.squeezeAxis, .squeezeAxis):
+      return true
+    case (let .index(i1), let .index(i2)): return i1 == i2
+    case (let .range(r1, s1), let .range(r2, s2)): return r1 == r2 && s1 == s2
+    case (let .closedRange(r1, s1), let .closedRange(r2, s2)):
+      return r1 == r2 && s1 == s2
+    case (let .partialRangeFrom(r1, s1), let .partialRangeFrom(r2, s2)):
+      return r1.lowerBound == r2.lowerBound && s1 == s2
+    case (let .partialRangeUpTo(r1, s1), let .partialRangeUpTo(r2, s2)):
+      return r1.upperBound == r2.upperBound && s1 == s2
+    case (let .partialRangeThrough(r1, s1), let .partialRangeThrough(r2, s2)):
+      return r1.upperBound == r2.upperBound && s1 == s2
+    default: return false
+    }
+  }
+}
+
+public let ellipsis: TensorRange = .ellipsis
+public let newAxis: TensorRange = .newAxis
+public let squeezeAxis: TensorRange = .squeezeAxis
+
+public func strided(_ value: Range<Int32>, by stride: Int32) -> TensorRange {
+  return .range(value, stride: stride)
+}
+
+public func strided(
+  _ value: ClosedRange<Int32>,
+  by stride: Int32
+) -> TensorRange {
+  return .closedRange(value, stride: stride)
+}
+
+public func strided(
+  _ value: PartialRangeFrom<Int32>,
+  by stride: Int32
+) -> TensorRange {
+  return .partialRangeFrom(value, stride: stride)
+}
+
+public func strided(
+  _ value: PartialRangeUpTo<Int32>,
+  by stride: Int32
+) -> TensorRange {
+  return .partialRangeUpTo(value, stride: stride)
+}
+
+public func strided(
+  _ value: PartialRangeThrough<Int32>,
+  by stride: Int32
+) -> TensorRange {
+  return .partialRangeThrough(value, stride: stride)
+}
+
+public protocol TensorRangeExpression {
+  var tensorRange: TensorRange { get }
+}
+
+// TODO: Cannot extend non-nominal type 'UnboundedRange'.
+// extension UnboundedRange : TensorRangeExpression {
```

Comment:
This is because the whole `UnboundedRange` thing is kind of a hack in Swift. There's no nullary operators, so the standard library just takes a reference to a binary operator function ([example](https://developer.apple.com/documentation/swift/array/3017543-subscript)).

Reply:
Yes I noticed this. It also means that currently we can't really support `...` as notation for ellipses inside lists of ranges (e.g., `tensor[1..<2, ..., 0...4]`). Is that right?

Reply:
Yeah right, since functions can't conform to protocols.

Reply:
It doesn't really make sense to support a subscript with a single ellipsis argument because that's equivalent to not slicing at all.

Reply:
Let's just require the user to do `TensorRange.ellipsis` for now.